{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0aa671ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime tools import --type 'SampleData[PairedEndSequencesWithQuality]' --input-path /Users/madibaltagulov/Documents/16s-Madi/TPT_diet-selected/TPT_manifest.tsv  --output-path paired-end-demux.qza --input-format PairedEndFastqManifestPhred33V2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f681d34e",
   "metadata": {},
   "source": [
    "Import sequencing files through the premade manifest file and respective. Create artifact file of quality scores for positions in our sequence for all reads. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e55610a",
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime demux summarize --i-data paired-end-demux.qza --o-visualization paired-end-demux.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "400e9249",
   "metadata": {},
   "source": [
    "Summarize and visualize the artifacts in a qzv file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20d2eea4",
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime dada2 denoise-paired --i-demultiplexed-seqs paired-end-demux.qza --p-trunc-len-f 150 --p-trunc-len-r 150 --p-trim-left-f 20 --p-trim-left-r 20 --o-representative-sequences dada2-rep-seqs.qza --o-table table.qza --o-denoising-stats stats-dada2.qza --p-n-threads 8\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a371b424",
   "metadata": {},
   "source": [
    "denoised the sequences we have, essentially trimming at bases that we decided have a low quality score (<25qc). 'trunc-len' is for 3' end bases, 'trim-left' is for 5' end bases. -f is for forward reads, -r for reverse reads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e475767f",
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime metadata tabulate --m-input-file stats-dada2.qza --o-visualization dada2-stats.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e87bd25",
   "metadata": {},
   "source": [
    "Makes a qzv file that makes the metadata and visualizing it in a table with denoised data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da23d471",
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime feature-table summarize --i-table table.qza --o-visualization table.qzv --m-sample-metadata-file /Users/madibaltagulov/Documents/16s-Madi/TPT_diet-selected/TPT_metadata.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22c042d9",
   "metadata": {},
   "source": [
    "This summarizes the feature table and pairs the metadata treatment groups with our sequences. A feature is a unique read (in V3-4 16s, that would be an Amplicon Sequence Variant (ASV))."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "567da151",
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime tools view 'insert file name'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35c8209a",
   "metadata": {},
   "source": [
    "allows to output any qzv file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d968a24a",
   "metadata": {},
   "outputs": [],
   "source": [
    "conda activate qiime2-2023.2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f4f822e",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee7844ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd 'file path'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0b2a94d",
   "metadata": {},
   "source": [
    "sets up the directory to work in"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0b42a7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime phylogeny align-to-tree-mafft-fasttree --i-sequences dada2-rep-seqs.qza --o-alignment aligned-rep-seqs.qza --o-masked-alignment masked-aligned-rep-seqs.qza --o-tree unrooted-tree.qza --o-rooted-tree rooted-tree.qza\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98a1e526",
   "metadata": {},
   "source": [
    "Creating a phylogeny tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fb60c03",
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime diversity alpha-rarefaction --i-table table.qza --i-phylogeny rooted-tree.qza --p-max-depth 27781 --m-metadata-file /Users/madibaltagulov/Documents/16s-Madi/TPT_diet-selected/TPT_metadata.txt --o-visualization alpha-rarefaction.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24d55ad8",
   "metadata": {},
   "source": [
    "Performing rarefaction (finding what minimum number of features/sample is appropriate to be representative of maximum amount of variation in each sample). This number should be sourced from the feature table qzv. Rarefaction was successful if every alpha diversity curve flatlines."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b311c55",
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime diversity core-metrics-phylogenetic --i-phylogeny rooted-tree.qza --i-table table.qza --p-sampling-depth 27781 --m-metadata-file /Users/madibaltagulov/Documents/16s-Madi/TPT_diet-selected/TPT_metadata.txt --output-dir core-metrics-results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2332d367",
   "metadata": {},
   "source": [
    "Performing an array of alpha-diversity and beta-diversity tests on our reads, coupled with metadata categories and labels. Ouputs into a new folder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6bcb7a7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime diversity alpha-group-significance --i-alpha-diversity core-metrics-results/shannon_vector.qza --m-metadata-file /Users/madibaltagulov/Documents/16s-Madi/TPT_diet-selected/TPT_metadata.txt --o-visualization core-metrics-results/shannon-vector.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf32bcf6",
   "metadata": {},
   "source": [
    "An example of how to output an alpha-diversity vector visualization. Also performs a Kruskal-Wallis (on each pair of categories) test to see if the differences btw metadata categories are significant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0923c61",
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime diversity beta-group-significance --i-distance-matrix core-metrics-results/unweighted_unifrac_distance_matrix.qza --m-metadata-file /Users/madibaltagulov/Documents/16s-Madi/TPT_diet-selected/TPT_metadata.txt --m-metadata-column Diet --o-visualization core-metrics-results/unweighted-unifrac-distance-matrix.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2c44c12",
   "metadata": {},
   "source": [
    "An example of how to output a beta-diversity distance matrix visualization. Also performs PERMANOVA test."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6326f0e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime emperor plot --i-pcoa core-metrics-results/bray_curtis_pcoa_results.qza --m-metadata-file /Users/madibaltagulov/Documents/16s-Madi/TPT_diet-selected/TPT_metadata.txt --o-visualization core-metrics-results/bray-curtis-distance-matrix.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f536008c",
   "metadata": {},
   "source": [
    "An example of how to output an emperor plot distance matrix (Principal Coordinate Analysis). It is a summation of beta diversity analyses of our data that is plotted on several axes. Percentage labels indicate the amount of variance explained by an axis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85815470",
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime feature-classifier classify-sklearn --i-classifier /Users/madibaltagulov/Documents/16s-Madi/TPT_diet-selected/16S_reads/2022.10.backbone.full-length.nb.qza --i-reads dada2-rep-seqs.qza --o-classification taxonomy.qza\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddebd2a8",
   "metadata": {},
   "source": [
    "Creating an actual taxonomy table using our reads and aligning them with Greenhils or Silva classifiers (database of bacteria genomes)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afc4fd93",
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime metadata tabulate --m-input-file taxonomy.qza --o-visualization taxonomy.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "072614cb",
   "metadata": {},
   "source": [
    "Visualizing the taxonomy table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad51725e",
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime taxa barplot --i-table table.qza --i-taxonomy taxonomy.qza --m-metadata-file /Users/madibaltagulov/Documents/16s-Madi/TPT_diet-selected/TPT_metadata.txt --o-visualization taxa-bar-plots.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a7d1325",
   "metadata": {},
   "source": [
    "Creating a bar plot that contains our taxonomy table, grouped by metadata categories. The bar plot is customizable by taxa level, treatment, frequence. It also demonstrated relative abundance. In 16S, we look at family or genus level usually."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8989a89",
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime taxa collapse --i-table table.qza --i-taxonomy taxonomy.qza --p-level 6 --o-collapsed-table genus-table.qza\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a6e595d",
   "metadata": {},
   "source": [
    "Collapsing our taxonomy level to a specific taxonomic level (in this example it's genus). There maybe be 0 values, as well as reads that are not determined down to the genus level."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a2cc250",
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime composition add-pseudocount --i-table genus-table.qza --o-composition-table genus_comp_table.qza\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b75cdfaf",
   "metadata": {},
   "source": [
    "Adds a pseudocount, transforming 0 values in the table to 1, as well as adding 1 to every value."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "438c3f3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime composition ancom --i-table genus_comp_table.qza --m-metadata-file /Users/madibaltagulov/Documents/16s-Madi/TPT_diet-selected/TPT_metadata.txt --m-metadata-column Diet --o-visualization genus-ancom-Diet.qzv\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a7382ca",
   "metadata": {},
   "source": [
    "Runs ANCOM on our data. ANCOM is a PERMANOVA-like test, which is gonna test what groups/taxa are significantly different among our diets (3 categories, 1 variable, non-zero values)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78b90045",
   "metadata": {},
   "outputs": [],
   "source": [
    "jupyter notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7093bb6",
   "metadata": {},
   "source": [
    "Code to open this through terminal."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
